File: sol/3318/3318.py
import heapq as hq

class Solution:
    def findXSum(self, nums: List[int], k: int, x: int) -> List[int]:
        n = len(nums)
        res = [0] * (n - k + 1)
        freq = dict()
        for z in nums[:k - 1]:
            freq[z] = freq.get(z, 0) + 1
        for i in range(n - k + 1):
            z = nums[i + k - 1]
            freq[z] = freq.get(z, 0) + 1
            h = []
            for z, f in freq.items():
                if f == 0:
                    continue
                hq.heappush(h, (f, z))
                if len(h) > x:
                    hq.heappop(h)
            res[i] = sum(z * f for f, z in h)
            freq[nums[i]] -= 1
        return res

File: sol/3318/3318.cpp
struct Solution {
    vector<int> findXSum(vector<int>& nums, int k, int x) {
        int n = nums.size();
        vector<int> res(n - k + 1, 0);
        unordered_map<int, int> freq;
        for (int i = 0; i < k - 1; ++i){
            ++freq[nums[i]];
        }
        for (int i = 0; i < n - k + 1; ++i){
            ++freq[nums[i + k - 1]];
            vector<pair<int, int>> h;
            for (auto [z, f] : freq){
                if (f == 0){
                    continue;
                }
                if (h.size() <= x){
                    h.push_back({-f, -z});
                } else {
                    h.back() = {-f, -z};
                }
                ranges::push_heap(h);
                if (h.size() > x){
                    ranges::pop_heap(h);
                }
            }
            int l = min(static_cast<size_t>(x), h.size());
            for (int j = 0; j < l; ++j){
                auto [f, z] = h[j];
                res[i] += z * f;
            }
            --freq[nums[i]];
        }
        return res;        
    }
};
